libvast/src/plugin.cpp

```diff
@@ -380,6 +382,55 @@ store_plugin::make_store(system::accountant_actor accountant,
                                                  std::move(path), name());
 }
 
+// -- loader plugin -------------------------------------------------------------
+
+auto stdin_loader_plugin::make_loader(options, const bool_operator*) const
+  -> caf::expected<input_loader> {
+  return []() -> generator<chunk_ptr> {
+    auto in_buf = detail::fdinbuf(STDIN_FILENO, max_chunk_size);
+    in_buf.read_timeout() = read_timeout;
+    auto chunk_str = std::string{};
+    chunk_str.reserve(max_chunk_size);
+    auto current_char = detail::fdinbuf::int_type{};
+    while (current_char != detail::fdinbuf::traits_type::eof()) {
+      auto current_char = in_buf.sbumpc();
+      if (current_char != detail::fdinbuf::traits_type::eof()) {
+        chunk_str += static_cast<char>(current_char);
+      } else if (in_buf.timed_out() && chunk_str.empty()) {
+        co_yield chunk::make_empty();
+        continue;
+      } else if (!in_buf.timed_out()) {
+        if (chunk_str.empty()) {
+          break;
+        }
+      }
```

Comment:
I think it's not ideal to check for the timeout for every single byte read, this seems like a lot of overhead to me. Can we do this in the outer loop?

Reply:
I specialized the timeout check to only trigger when we reach `eof()` since `fdinbuf` returns an `eof()` upon reaching a timeout. https://github.com/tenzir/vast/pull/2935/commits/ea29892b310fee9a14ce51876008f8093b946235

Reply:
There are probably no error scenarios for this specific loader, but I can imagine for example a TCP loader that would stop with a meaningful error when the handshake fails. Maybe we can just postpone this discussion until the next loader, and for now just blindly take a `operator_control_plane& ctrl`.